Docker authorization plug-in infrastructure

liron-l opened the pull request:
REMARK: Code is intended to be complementary to the docker (authZ discussion) [https://github.com/docker/docker/issues/14674]. There are several aspects of the proposed code which require additional work and testing. Please use this commit as a low level design proposal.
# 

Docker authorization plug-in infrastructure enables extending the functionality of the Docker daemon with respect to user authorization.
The infrastructure enables registering a set of external authorization
plug-in. Each plug-in receives information about the user and the
request and decides whether to allow or deny the request. Only in case
all plug-ins allow accessing the resource the access is granted.

Each plug-in operates as a separate service, and registers with Docker
through general (plug-ins API)
[https://blog.docker.com/2015/06/extending-docker-with-plugins/]. No
Docker daemon recompilation is required in order to add / remove an
authentication plug-in. Each plug-in is notified twice for each
operation: 1) before the operation is performed and, 2) before the
response is returned to the client. The plug-ins can modify the response
that is returned to the client.

The authorization depends on the authorization effort that takes place
in parallel [https://github.com/docker/docker/issues/13697].

This is the official issue of the authorization effort:
https://github.com/docker/docker/issues/14674

(Here)[https://github.com/rhatdan/docker-rbac] you can find an open
document that discusses a default RBAC plug-in for Docker.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14674-docker-authz" git@github.com:twistlock/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
cc @icecrime @crosbymichael  @NathanMcCauley @diogomonica perhaps you can have a look at this? General API description can be found [in this document](https://github.com/twistlock/docker/blob/14674-docker-authz/pkg/authorization/api.md)

NathanMcCauley commented on pkg/authorization/authz.go:
```diff
@@ -0,0 +1,130 @@
+		if err != nil {
+			return err
+		}
+		a.plugins = append(a.plugins, p)
```
Is it necessary to parse and build this on every request? Could instead the `Ctx` struct be built with a set of `authZPlugins` of type `[]*plugins.Plugin`? Or better yet does it make sense to create a typed plugin that could be plugins.AuthZPlugin`?

NathanMcCauley commented:
Should there any limitations to what the Response rewriters can and cannot do? It looks like the proposal allows arbitrary modifications? I'm concerned that'll mean that we end up with accidental dependencies on the structure of the Docker Remote API request/response structure as 3rd party authz plugins now depend on being able to directly parse the response bodies.

NathanMcCauley commented:
How will this handle cases where data is streamed live back to the client, as in the case of the `/logs` endpoint. How will it deal with `attach`?

NathanMcCauley commented on pkg/authorization/api.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// Request is the expect object that returns from the authZ plugin
+type Response struct {
```
In #14674 @dimastopel mentions "Whether the next plug-in should be called or not for this specific notification (default is yes)." I don't see that represented in the API? Has that notion been left out of the proposal?

dimastopel replied:
We have decided to go with a simpler approach as the first step and tune / optimize as we go. Current approach is to simply AND between all the plugins and only proceed if the access is approved by all. This can definitely be revisited.

liron-l replied:
@NathanMcCauley, thanks for you comments.
Re /logs and /attache, the authentication framework should only be called on the HTTP requests. Once the http connection is hijacked, no additional authorization should to be perform. This pattern enables full authorization capabilities to docker as described in the current requirements (for framework and specific RBAC plugin).  The only change in current code is to replace the response recorder with concrete object that implements hijack.

Re response rewrites, since the api version is exposed to the plugin, I think it should be the plugin responsibility to determine whether it can or cannot modify the response or which parts can be changed.  As far as I know, the docker daemon api is backward compatible, so plugins are essentially equivalent to regular client communicating with the api. 

I totally agree with other comments, please let me know your thoughts on the above items, and I will modify the code appropriately.

NathanMcCauley commented on pkg/authorization/authz.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// Write headers and body
+	w.WriteHeader(rr.Code)
```
Should authz plugins be able to modify headers as well?

liron-l replied:
Yes, based on the feedback we've modified the current api to enable both body, header and status code manipulation (see updated md or api.go).

diogomonica commented on api/server/server.go:
```diff
@@ -17,18 +17,21 @@ import (
+	Version      string
+	SocketGroup  string
+	TLSConfig    *tls.Config
+	AuthZPlugins []string
```
My expectation from the variable name is for this to be a type.

liron-l replied:
Thanks. I've replaced it in the go code except from daemon config.
I find it more clear to use ---authz-plugins=xxx than --authz-plugin-names=xxx, what do you think?

diogomonica commented on api/server/server.go:
```diff
@@ -289,9 +292,44 @@ func makeHTTPHandler(logging bool, localMethod string, localRoute string, handle
+		if len(authZPlugins) > 0 {
+
+			// User and UserAuthNMethod are taken from AuthN plugins
+			// Proposal fro implemenation https://github.com/docker/docker/pull/13994
```
small nit: this comment has at least 2 typos.

liron-l replied:
Thanks and sorry

liron-l commented:
Added unit tests and fixed implementation according to comments.

liron-l commented:
@icecrime @crosbymichael @NathanMcCauley @diogomonica 
Added detailed design document for the authorization feature [here] (https://docs.google.com/a/twistlock.io/document/d/1WH7IPfri8TwxhyL1fukNA2rtfib7J8eVPvhobyDID9c/edit?usp=sharing).

jfrazelle commented:
How does this work w PAM, it would be nice if it was easy for people to understand and familiar the design that way

liron-l commented:
Thanks for the review @diogomonica, @NathanMcCauley, @endophage,

We've updated the design doc and answered all comments, looking forward to discuss further and agree on the right approach for plugin authorization.
I expect that in the majority of scenarios, plugins will not be required to perform any additional queries to the daemon. However, we should find a proper mechanism for such scenarios (e.g., for plugins that need to scan the content of an image to decide whether permission should be granted or not).
Aggregating the feedback, I'd like to propose three possible solutions:

1. Capability based solution -
    In the daemon level, each plugin will be configured with a dedicated policy for the command it can invoke (e.g., pull, inspect) 
2. Authorization policy based solution - 
    Admin sets up a dedicated user for the plugin and defines a policy for this user. The plugin will be authenticated and authorized as a regular user.
3. Sending required metadata - 
For example, send all image info for container run command. 
I think we can rule out this solution since this can lead to sending a large chunk of redundant data. For example, assume that some flow requires examining the image history, which leads to sending the entire history for each command related to images, a scenario which we would probably like to avoid.

I personally think both solution 1 and 2 are plausible, but the former (1) will require much more logic in the daemon core, and will have to be maintained anytime new commands are added (which can have some maintenance pain).

What do you think?

lioryanko commented:
@jfrazelle PAM is one of the mechanisms which can be used for authentication by plugins which will be built on top of the authentication framework (https://github.com/docker/docker/issues/13697) that is currently under development. Authorization plugins will depend upon the authentication data to achieve access control with greater granularity than the current dichotomy of full access or no access.
We'd love you to ask further questions and give suggestions regarding PAM and authentication in general at https://github.com/docker/docker/issues/13697 or straight at [the design and UX doc](https://docs.google.com/document/d/1lQr-9wnhd7yI7t8rF7cnB6hGGUZoYk6wvMO2ePLcx68/edit)

mheon commented:
@liron-l Looking at the options you've laid out, I would tend to prefer the third from an architectural perspective, as it does not require additional requests to the public API. However, I can certainly understand how this might massively increase overhead, particularly in cases dealing with images. Option one would be a good alternative, assuming we have a reliable way of authenticating plugins (perhaps they could declare a public key in the plugin specification?). I really don't like the second option - it seems the sort of thing where it would be very easy for a user to shoot themselves in the foot if they had multiple authorization plugins.

As a possible alternative, perhaps plugins could tell the daemon they required more information to make a decision, and the daemon could reissue the authorization request with additional information on the image or container in question. This would significantly increase the complexity of authorization responded and the complexity of the daemon code, but would remove the need for additional requests from the authorization plugin to the public API.

thaJeztah commented:
ping @diogomonica, @NathanMcCauley, @jfrazelle PTAL (https://github.com/docker/docker/pull/15365#issuecomment-133906874)

calavera commented:
I'd rather avoid premature optimizations. That will allow us to get this in sooner rather than later. If we don't need cross references with the daemon, let's not add them from now.

From a design point of view, this change looks good to me, but I'll wait for other maintainers to confirm.
From a code point of view, there have been a lot of changes in the server package that conflict with this. We can probably reduce the conflicts transforming this code in a middleware like we've done with other pre processing code in the server:

https://github.com/docker/docker/blob/master/api/server/middleware.go

jfrazelle commented:
I agree with @calavera and I think using middleware for this would be great.

liron-l commented:
@calavera, @jfrazelle thanks for the feedback and sounds good and much more appropriate moving the authZ plugin code to the middleware layer.
Any additional feedback about design/implemenation we should consider?

calavera commented:
@NathanMcCauley any concerns about the design?

Commit:
Adding authorization subsystem documentation

Signed-off-by: Dima Stopel <dima@twistlock.com>

calavera commented on api/server/server.go:
```diff
@@ -289,9 +291,40 @@ func makeHTTPHandler(logging bool, localMethod string, localRoute string, handle
+				return
+			}
+
+			if err := ctx.AuthZResponse(version, rw, r, mux.Vars(r)); err != nil {
```
I'm having issues finding a use case for this `AuthZResponse`. Why would we want to let plugins to modify the api response after granting access in the request?

calavera commented:
This is what I think we should do to move this forwards, I'll be happy to assist you in whatever I can, but I think it would be best if you could complete the patch:

**1-** Move the auth check to a middleware what will be injected only when there are authz plugins set, like we do with the debug middleware:

https://github.com/docker/docker/blob/master/api/server/middleware.go#L142-L147

**2-** After talking with @NathanMcCauley, I think we should remove the response management for two reasons:

- It's very unexpected to allow third party applications to manipulate the response output, and very dangerous, from my point of view.
- We won't probably be able to modify the response status code. Go guarranties that it's set only once. When the `AuthzResponse` function is called, the status might have been set already. Trying to ensure the code is not set before calling that function can be a mayor pain for the first version of this feature.

**3-** Integration tests, lots of them :smile_cat: We can create a mock plugin like we already have for volumes and networking and start a daemon for testing with it:

https://github.com/docker/docker/blob/master/integration-cli/docker_cli_start_volume_driver_unix_test.go#L54-L175

Commit:
Docker authorization plug-in infrastructure enables extending the functionality of the Docker daemon with respect to user authorization. The infrastructure enables registering a set of external authorization plug-in. Each plug-in receives information about the user and the request and decides whether to allow or deny the request. Only in case all plug-ins allow accessing the resource the access is granted.

Each plug-in operates as a separate service, and registers with Docker
through general (plug-ins API)
[https://blog.docker.com/2015/06/extending-docker-with-plugins/]. No
Docker daemon recompilation is required in order to add / remove an
authentication plug-in. Each plug-in is notified twice for each
operation: 1) before the operation is performed and, 2) before the
response is returned to the client. The plug-ins can modify the response
that is returned to the client.

The authorization depends on the authorization effort that takes place
in parallel [https://github.com/docker/docker/issues/13697].

This is the official issue of the authorization effort:
https://github.com/docker/docker/issues/14674

(Here)[https://github.com/rhatdan/docker-rbac] you can find an open
document that discusses a default RBAC plug-in for Docker.

Signed-off-by: Liron Levin <liron@twistlock.com>
Added container create flow test and extended the verification for ps

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14674-docker-authz" git@github.com:twistlock/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

cyphar commented:
@calavera 

>It's very unexpected to allow third party applications to manipulate the response output,

Well, what if the AuthZ plugin needs to keep track of tokens or other such things? Maybe allowing the plugin to set a particular Header within the response would be better (such as it returning the header value to pass to the response)?

liron-l commented:
@calavera thanks for the review, I've made the following changes in the original code:
1. Moved the authz handling to middleware 
2. Removed ability to modify the response from authz plugin. 
The middleware can still override the response returned from docker daemon (on deny scenario), but the implementation handles the GO limitation by wrapping the HTTP response (this is also tested in integration tests and UT). If you still think this is wrong, we can remove the ability to deny the response flow and just send the data. The response flow is essential since plugins may need to know whether the flow completed successfully. WDYT?
3. I've added unit tests and integration tests with mock plugins.

calavera commented:
Thanks @liron-l. I agree with the design decisions. I'm moving this to code review so we can move it forward.

calavera commented on integration-cli/docker_cli_authz_unix_test.go:
```diff
@@ -0,0 +1,166 @@
+	c.Assert(res, check.Matches, fmt.Sprintf(".*%s\n", unauthorizedMessage))
+}
+
+func (s *DockerAuthzSuite) TestAuthZPluginDenyResponse(c *check.C) {
```
can we test here that the normal PS output is not flushed in the output?

calavera replied:
@liron-l can you make this test to assert that the normal PS output is not in the output? it would make the test stronger to changes.

liron-l replied:
Thanks @calavera this makes sense. I think the cleanest and maintainable approach is to fully compare the err response. I've modified the code appropriately, let me know what you think.

runcom commented on pkg/authorization/authz.go:
```diff
@@ -0,0 +1,158 @@
+// AuthZRequest authorized the request to the docker daemon using authZ plugins
+func (a *Ctx) AuthZRequest(w http.ResponseWriter, r *http.Request) (err error) {
+
+	if err != nil {
```
is there something missing above here?

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14674-docker-authz" git@github.com:twistlock/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

liron-l commented:
Thanks for the review @calavera and @runcom. 
I've extended the existing integration tests with additional commands and output validations.

calavera commented:
@liron-l it looks like you latest changes broke some unit tests, can you take a look?

calavera commented:
Thanks for adding the documentation in a separated commit, btw.

liron-l commented:
Thanks @calavera, I don't think the failures were related to my change. 
I've rebased with the latest code and everything looks good.

calavera commented:
LGTM

jfrazelle commented:
Let's do this LGTM 😁

cpuguy83 commented on pkg/authorization/authz.go:
```diff
@@ -0,0 +1,154 @@
+func (a *Ctx) AuthZRequest(w http.ResponseWriter, r *http.Request) (err error) {
+
+	var drainedBody io.ReadCloser
+	drainedBody, r.Body, err = drainBody(r.Body)
```
Is there a real need to send the body of the request? This could be quite large in some cases.

liron-l commented:
Thanks for the review guys. 
@cpuguy83 - The authz plugins receive the data only when the Content-Type is either application/json or *text/*. This ensures only the content of the docker remote API is passed to the plugins.
For commands that involve the hijacking of the HTTP connection (HTTP Upgrade), such as `exec`, the authorization plugins will only be called for the initial HTTP requests, once the commands are approved, authorization will not be applied to the rest of the flow. Specifically, the streaming data will not be passed to the authorization plugins.

@thaJeztah, the command line only supports passing the authz-plugins parameter multiple times (same as other list parameters, e.g., insecure-registry, dns-opt, etc...). I've kept the parameter name 'plugins' plural to indicate multiple plugins are supported and fixed the submitted documentation accordingly.

cpuguy83 commented:
@liron-l It looks like we're draining the body into a buffer no matter what, and only sending to the client if it's text/ or application/json.

liron-l commented:
Good catch @cpuguy83. I've refactor the code according to your comment. 
I've also added an explicit validation for body length in the integration tests.

cpuguy83 commented:
Honestly I'm not sure what the benefit is of passing the request (or the response) body to the auth plugin.
In addition, `docker login` will indeed pass credentials to the auth plugin.

liron-l commented:
Adding the body enables the authz plugin to use the entire docker REST API. 
For example, allow/deny [creating containers] (https://docs.docker.com/engine/reference/api/docker_remote_api_v1.21/#create-a-container) with specific tags. 
This is equivalent to Kubernetes authorization approach that uses namespaces and resources. 

Re login, the authorization plugin explicitly removes the authorization header. Does this makes sense?

cpuguy83 commented:
@liron-l The specific case with `docker login` is the client sends the registry credentials in the POST body to the engine daemon.

liron-l commented:
Makes sense @cpuguy83, I've removed the the body from calls to the the /auth endpoint and also excluded the following headers: X-Registry-Config, X-Registry-Auth and Authorization. 
I've also added validations in the integration tests.

calavera commented:
This needs a rebase.

@cpuguy83 can you take a look at the changes?

icecrime commented:
Moving back to code review until we have @cpuguy83 approval.

cpuguy83 commented:
Changes LGTM

icecrime commented:
Well.

cpuguy83 commented:
:)

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,120 @@
+
+# Access authorization
+
+By default Docker’s authorization model is all or nothing. Anyone who has access to the Docker daemon has the ability to run any Docker command. Authorization plugins enable creating granular access policies for managing access to Docker daemon. 
```
I wonder if we should refer to the docker *daemon* or the docker remote API; "Anyone who has access to the Docker remote API.."

The last sentence is quite long and a bit confusing; perhaps

```
Authorization plugins enable you to create granular access policies to the Docker remote API.
```

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,120 @@
+
+## Basic principles
+
+* The authorization sub-system is based on Docker's [Plugin API](http://docs.docker.com/engine/extend/plugin_api), so that anyone could add / develop an authorization plug-in. It is possible to add a plug-in to a deployed Docker daemon without the need to rebuild the daemon. Daemon restart is required.
```
I think "so that anyone could add / develop an authorization plug-in" may be redundant

thaJeztah replied:
Also, can you change the link to `plugin_api.md`; the docs generator should convert that to the right URL (you can test that by running `make docs`

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,120 @@
+
+* The authorization sub-system is based on Docker's [Plugin API](http://docs.docker.com/engine/extend/plugin_api), so that anyone could add / develop an authorization plug-in. It is possible to add a plug-in to a deployed Docker daemon without the need to rebuild the daemon. Daemon restart is required.
+
+* The authorization sub-system enables approving or denying requests to the Docker daemon based on the current user and command context, where the authentication context contains all user details and the authentication method, and the command context contains all the relevant request data.
```
```
The authorization sub-system enables approving or denying requests,
based on the "authentication" and "command" context provided in the
request. The "authentication" context contains all user details, and
the "command" context contains all relevant request data.
```

Trying to avoid "current user" here, because afaik, the API is stateless, so there's no "current" user,
only the information that is provided in the request (?)

dimastopel replied:
Removed the "current" but left the user and command context. The user and the command are part of the request context (not state of the API). Please let me know if you prefer to rephrase.

thaJeztah replied:
probably better to use @moxiegirl's example; I think she did a whole rewrite of the docs that you can copy/paste

dimastopel replied:
Yes, agree

Dima Stopel | e: dima at twistlock dot com | p: +972.545.959.320 |
gpg: 71FF 77BA 8989 2CCD



On Mon, Dec 7, 2015 at 3:27 PM, Sebastiaan van Stijn <
notifications@github.com> wrote:

> In docs/extend/authorization.md
> <https://github.com/docker/docker/pull/15365#discussion_r46819718>:
>
> > +[menu.main]
> > +parent = "mn_extend"
> > +weight = -1
> > ++++
> > +<![end-metadata]-->
> > +
> > +
> > +# Access authorization
> > +
> > +By default Docker’s authorization model is all or nothing. Anyone who has access to the Docker daemon has the ability to run any Docker command. Authorization plugins enable creating granular access policies for managing access to Docker daemon.
> > +
> > +## Basic principles
> > +
> > +* The authorization sub-system is based on Docker's [Plugin API](http://docs.docker.com/engine/extend/plugin_api), so that anyone could add / develop an authorization plug-in. It is possible to add a plug-in to a deployed Docker daemon without the need to rebuild the daemon. Daemon restart is required.
> > +
> > +* The authorization sub-system enables approving or denying requests to the Docker daemon based on the current user and command context, where the authentication context contains all user details and the authentication method, and the command context contains all the relevant request data.
>
> probably better to use @moxiegirl <https://github.com/moxiegirl>'s
> example; I think she did a whole rewrite of the docs that you can copy/paste
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/15365/files#r46819718>.
>

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,120 @@
+
+* The authorization sub-system enables approving or denying requests to the Docker daemon based on the current user and command context, where the authentication context contains all user details and the authentication method, and the command context contains all the relevant request data.
+
+* The authorization sub-system enables approving or denying the response from the Docker daemon based on the current user and the command context.
```
Same here, also lots of duplication with the previous bullet; at first I thought they were the same, but this is for the *response*.

thaJeztah commented:
@liron-l I had a quick chat with @moxiegirl, who is one of our technical writers, and she'll be looking at the documentation later on. We think this is an important feature, so it needs the right attention for the documentation.

I'll continue reviewing after she's had a first pass on the documentation, thanks for being patient with us; it's a really cool feature!

moxiegirl commented on daemon/config.go:
```diff
@@ -54,6 +55,7 @@ type CommonConfig struct {
 // from the command-line.
 func (config *Config) InstallCommonFlags(cmd *flag.FlagSet, usageFn func(string) string) {
 	cmd.Var(opts.NewListOptsRef(&config.GraphOptions, nil), []string{"-storage-opt"}, usageFn("Set storage driver options"))
+	cmd.Var(opts.NewListOptsRef(&config.AuthZPlugins, nil), []string{"-authz-plugins"}, usageFn("List of authorization plugins by order of evaluation"))
```
Evaluation is what is done to something What I think is  really meant is that I can have multiple plugins installed.  A user seeking authorization passes through each one and is evaluated.   Short of writing all that you can change the wording a bit and still convey that. _evaluator n. Agent noun of evaluate; one who evaluates._  

```markdown
List authorization plugins in order from first evaluator to last.
```

or

```markdown
List authorization plugins in order from first authorization check to last.
```

moxiegirl commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,120 @@
+<!--[metadata]>
```
@liron-l  thank you for the contribution. As @thaJeztah points out, there are several redundancies in this information. Also, structurally, I think there are opportunities to help the reader out by tightening the language used.  A couple of questions come to mind:

* The text mentions "authentication methods" are passed to the plugin. This implies there is some set of possible methods a plugin can expect. If there are, might a developer want to know what these are?

* It would be nice to have a diagram of the flow here to help the reader visualize.

* Is there a sample plugin we can point them to in our code or elsewhere?

Below I've included the complete text the incorporates all my comments. You can reach at the gist indicated.  


---> https://gist.github.com/moxiegirl/852dbab3b392c28e29b2

<!--[metadata]>
+++
title = "Access authorization plugin"
description = "How to create authorization plugins to manage access control to your Docker daemon."
keywords = ["security, authorization, authentication, docker, documentation, plugin, extend"]
[menu.main]
parent = "mn_extend"
weight = -1
+++
<![end-metadata]-->


# Create an authorization plugin

Docker’s out-of-the-box authorization model is all or nothing. Any user with
permission to access to the Docker daemon can run any Docker client command. The
same is true for callers using Docker's remote API to contact the daemon. If you
require greater access control, you can create authorization plugins and add
them to your Docker daemon configuration. Using an authorization plugin, a
Docker administrator can configure granular access policies for managing access
to Docker daemon.

Anyone with the appropriate skills can develop an authorization plugin. These
skills, at their most basic, are knowledge of Docker, understanding of REST, and
sound programming knowledge. This document describes the architecture, state,
and methods information available to an authorization plugin developer.

## Basic principles

Docker's [plugin
infrastructure](http://docs.docker.com/engine/extend/plugin_api) enables
extending Docker by dynamically loading, removing and communicating with
third-party components using a generic API. The access authorization subsystem
was built using this mechanism.

Using this subsystem, you don't need to rebuild the Docker daemon to add an
authorization plugin.  You can add a plugin to a installed Docker daemon. You do
need to restart the Docker daemon to add a new plugin.

An authorization plugin approves or denies requests to the Docker daemon based
on both the current authentication context and the command context. The
authentication context contains all user details and the authentication method.
The command context contains all the relevant request data.

## Basic architecture

You are responsible for registering your plugin as part of the Docker daemon
startup. You can install multiple plugins and chain them together. This chain
can be ordered. Each request to the daemon passes in order through the chain.
Only when all the plugins grant access to the resource, is the access granted.

When an HTTP request is made to the Docker daemon through the CLI or via the
remote API, the authentication subsystem passes the request to the installed
authentication plugin(s). The request contains the user (caller) and command
context. The plugin is responsible for deciding whether to allow or deny the
request.

Each request sent to the plugin includes the authenticated user, the HTTP
headers, and the request/response body. Only the user name and the
authentication method used are passed to the plugin. Most importantly, no user
credentials or tokens are passed. Finally, not all request/response bodies
are sent to the authorization plugin. Only those request/response bodies where
the `Content-Type` is either `text/*` or `application/json` are sent.

For commands that can potentially the hijack the HTTP connection (`HTTP
Upgrade`), such as `exec`, the authorization plugin are only called for the
initial HTTP requests. Once the plugin approves the command, authorization is
not applied to the rest of the flow. Specifically, the streaming data is not
passed to the authorization plugins. For commands that return chunked HTTP
response, such as `logs` and `events`, only the HTTP request are sent to the
authorization plug-ins as well.

During request/response processing, some authorization plugins flows might
need to do additional queries to the Docker daemon. To complete such flows,
plugins can call the daemon API similar to a regular user. To enable these
additional queries, the plugin must provide the means for an administrator to
configure proper authentication and security policies.

## API schema

In addition to Docker's standard plugin registration method, each plugin should implement the following two methods:

* `/AuthzPlugin.AuthZReq` This authorize request method is called before the Docker daemon processes the client request.

* `/AuthzPlugin.AuthZRes` This authorize response method is called before the response is returned from Docker daemon to the client.


### Request authorization

Each plugin must support two request authorization messages formats, one from the daemon to the plugin and then from the plugin to the daemon. The tables below detail the content expected in each message.

#### Daemon -> Plugin

Name                   | Type              | Description
-----------------------|-------------------|-------------------------------------------------------
User                   | string            | The user identification
Authentication method  | string            | The authentication method used
Request method         | enum              | The HTTP method (GET/DELETE/POST)
Request URI            | string            | The HTTP request URI including API version (e.g., v.1.17/containers/json)
Request headers        | map[string]string | Request headers as key value pairs (without the authorization header)
Request body           | []byte            | Raw request body


#### Plugin -> Daemon

Name    | Type   | Description
--------|--------|----------------------------------------------------------------------------------
Allow   | bool   | Boolean value indicating whether the request is allowed or denied
Message | string | Authorization message (will be returned to the client in case the access is denied)

### Response authorization

The plugin must support two authorization messages formats, one from the daemon to the plugin and then from the plugin to the daemon. The tables below detail the content expected in each message.

#### Daemon -> Plugin


Name                    | Type              | Description
----------------------- |------------------ |----------------------------------------------------
User                    | string            | The user identification
Authentication method   | string            | The authentication method used
Request method          | string            | The HTTP method (GET/DELETE/POST)
Request URI             | string            | The http request URI including API version (e.g., v.1.17/containers/json)
Request headers         | map[string]string | Request headers as key value pairs (without the authorization header)
Request body            | []byte            | Raw request body
Response status code    | int               | Status code from the docker daemon
Response headers        | map[string]string | Response headers as key value pairs
Response body           | []byte            | Raw docker daemon response body


#### Plugin -> Daemon

Name    | Type   | Description
--------|--------|----------------------------------------------------------------------------------
Allow   | bool   | Boolean value indicating whether the response is allowed or denied
Message | string | Authorization message (will be returned to the client in case the access is denied)

## Docker client flows

To enable and configure the authorization plugin, the plugin developer must support the Docker client interactions detailed in this section.

### Setting up Docker daemon

Enable the authorization plugin with a dedicated command line flag in the
`--authz-plugins=PLUGIN_ID` format. The flag supplies a `PLUGIN_ID` value.
This value can be the plugin’s socket or a path to a specification file.

```
$ docker daemon --authz-plugins=plugin1 --auth-plugins=plugin2,...
```

Docker's authorization subsystem supports multiple `--authz-plugin` parameters.

### Calling authorized command (allow)

Your plugin must support calling the `allow` command to authorize a command. This call does not impact Docker's command line.

```
$ docker pull centos
...
f1b10cd84249: Pull complete
...
```

### Calling unauthorized command (deny)

Your plugin must support calling the `deny` command to report on the outcome of a plugin interaction. This call returns messages to Docker's command line informing the user of the outcome of each call.  

```
$ docker pull centos
…
Authorization failed. Pull command for user 'john_doe' is denied by authorization plugin 'ACME' with message ‘[ACME] User 'john_doe' is not allowed to perform the pull command’
```

Where multiple authorization plugins are installed, multiple messages are expected.

moxiegirl commented on integration-cli/docker_cli_help_test.go:
```diff
@@ -133,7 +133,7 @@ func (s *DockerSuite) TestHelpTextVerify(c *check.C) {
 			lines := strings.Split(out, "\n")
 			for _, line := range lines {
-				c.Assert(len(line), checker.LessOrEqualThan, 90, check.Commentf("Help for %q is too long:\n%s", cmd, line))
+				c.Assert(len(line), checker.LessOrEqualThan, 91, check.Commentf("Help for %q is too long:\n%s", cmd, line))
```
`("Help for %q is too long:\n%s" -- the limit is XX characters, cmd, line)`

moxiegirl commented on docs/reference/commandline/daemon.md:
```diff
@@ -581,6 +582,12 @@ The currently supported cluster store options are:
     private key is used as the client key for communication with the
     Key/Value store.
 
+## Access authorization
```
It isn't necessary to include all that information about how plugins are  built or where they are located in the middle of discussing how to use the option.  There is some detail that may be helpful.

```markdown
## Access authorization

Docker's access authorization can be extended by authorization plugins that your
organization can buy or build themselves. You can install one or more
authorization plugins when you start the Docker `daemon` using the
`--authz-plugins=PLUGIN_ID` option. T
```
```
$ docker daemon --authz-plugins=plugin1 --auth-plugins=plugin2,...
```
```markdown
The`PLUGIN_ID` value is either the plugin’s name or a path to its specification
file. The plugin's implementation determines whether you can specify a name or
path. Consult with your Docker administrator to get information about the
plugins available to you.

Once a plugin is installed, requests made to the `daemon` through the command
line or Docker's remote API are authorized or not by the plugin.  If you have
multiple plugins installed, at least one must authorize the request for it to
complete.

For information about how to create an authorization plugin, see [authorization
plugin](https://docs.docker.com/engine/extend/authorization/) section in the
Docker extend section of this documentation.
```

moxiegirl commented on pkg/authorization/api.md:
```diff
@@ -0,0 +1,83 @@
+# Docker Authorization Plug-in API
```
I think this should be moved to a new `/docs/extend/reference` directory.  Leaving it in the package makes it hard to find and not so great for SEO

dimastopel commented:
@cpuguy83 thanks!

@thaJeztah @moxiegirl thanks for the comments! I will start working on the documentation fixes in the next couple of days.

thaJeztah commented:
Thanks, @dimastopel, ping us when it's ready for another review :+1:

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14674-docker-authz" git@github.com:twistlock/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~6
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

Commit:
Fixing documentation according to comments by @moxiegirl and @thaJeztah

Signed-off-by: Dima Stopel <dima@twistlock.com>

dimastopel commented:
@moxiegirl thanks a lot for going through all the documentation! 

I've just pushed the fixes. 

Some notes:

1. I didn't add the specific authentication methods yet. Those will be finalized when the authentication work be completed (#17412). I prefer to add them then. 
2. I've added some sequence diagrams to authorization.md file. Placed those under images. Hope it's OK. 
3. Regarding the plugin code example, we have the one we use for test. Would you like to create a well commented skeleton of it and embed into documentation, or referencing it is enough? I can do both. 
https://github.com/twistlock/docker/blob/14674-docker-authz/integration-cli/docker_cli_authz_unix_test.go
4. @moxiegirl I didn't understand your comment regarding "including the limit" above. Can you please elaborate?

@thaJeztah @moxiegirl other than that I think I have addressed all the comments. Looking forward to hear your thoughts.

cpuguy83 commented:
@liron-l Looks like there was a bad rebase, also needs another rebase.

Commit:
Rebase from master

Signed-off-by: Liron Levin <liron@twistlock.com>

liron-l commented:
Thanks @cpuguy83. @dimastopel fixed the documentation comments and I did a final rebased from in a separate commit.

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+# Create an authorization plugin
+
+Docker’s out-of-the-box authorization model is all or nothing. Any user with
+permission to access to the Docker daemon can run any Docker client command. The
```
"to access to the" should probably be "to access the"

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+
+## Basic principles
+
+Docker's [plugin infrastructure](http://docs.docker.com/engine/extend/plugin_api) enables
```
can you change the URL to `plugin_api.md`? That way it'll work both on GitHub and in the generated documentation

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+## Basic principles
+
+Docker's [plugin infrastructure](http://docs.docker.com/engine/extend/plugin_api) enables
+extending Docker by dynamically loading, removing and communicating with
```
Should we use "dynamically loading"? IIUC, I have to specify which plugins to use through a daemon flag, so a daemon restart is required to enable/disable a plugin?

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+was built using this mechanism.
+
+Using this subsystem, you don't need to rebuild the Docker daemon to add an
+authorization plugin.  You can add a plugin to a installed Docker daemon. You do
```
Nit; "to a installed" -> "to an installed"

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+The command context contains all the relevant request data.
+
+Authorization plugins must follow the rules described in [Docker Plugin API](http://docs.docker.com/engine/extend/plugin_api/). 
+Each plugin must reside within directories described under the [plugin discovery](http://docs.docker.com/engine/extend/plugin_api/#plugin-discovery)
```
Can you change these URLs as well (`plugin_api.md`)?

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+context. The plugin is responsible for deciding whether to allow or deny the
+request.
+
+Below please find tow sequence diagrams describing the *allow* and *deny* 
```
"The sequence diagrams below depict an *allow* and *deny* authorization flow:"

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+
+![Bad Address](images/authz_allow.png)
+
+![Bad Address](images/authz_deny.png)
```
Can you update the alt-texts to match what the images show?

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+are sent to the authorization plugin. Only those request/response bodies where
+the `Content-Type` is either `text/*` or `application/json` are sent.
+
+For commands that can potentially the hijack the HTTP connection (`HTTP
```
s/the hijack/hijack/

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+the `Content-Type` is either `text/*` or `application/json` are sent.
+
+For commands that can potentially the hijack the HTTP connection (`HTTP
+Upgrade`), such as `exec`, the authorization plugin are only called for the
```
s/plugin/plugins/

thaJeztah replied:
or "is only called"

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+initial HTTP requests. Once the plugin approves the command, authorization is
+not applied to the rest of the flow. Specifically, the streaming data is not
+passed to the authorization plugins. For commands that return chunked HTTP
+response, such as `logs` and `events`, only the HTTP request are sent to the
```
requests (plural), or "is sent"

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+not applied to the rest of the flow. Specifically, the streaming data is not
+passed to the authorization plugins. For commands that return chunked HTTP
+response, such as `logs` and `events`, only the HTTP request are sent to the
+authorization plugins as well.
```
Remove "as well" here

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+response, such as `logs` and `events`, only the HTTP request are sent to the
+authorization plugins as well.
+
+During request/response processing, some authorization plugins flows might
```
s/plugins/plugin/

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+
+## Docker client flows
+
+To enable and configure the authorization plugin, the plugin developer must support the Docker client interactions detailed in this section.
```
Can you wrap this to 80 chars?

thaJeztah commented on docs/extend/authorization.md:
````diff
@@ -0,0 +1,243 @@
+This value can be the plugin’s socket or a path to a specification file.
+
+```
+$ docker daemon --authz-plugins=plugin1 --auth-plugins=plugin2,...
````
typo in the second flag; `auth-plugins` -> `authz-plugins`

thaJeztah commented on docs/extend/authorization.md:
````diff
@@ -0,0 +1,243 @@
+$ docker daemon --authz-plugins=plugin1 --auth-plugins=plugin2,...
+```
+
+Docker's authorization subsystem supports multiple `--authz-plugin` parameters.
````
Should be `--authz-plugins` (plural)

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+
+### Calling authorized command (allow)
+
+Your plugin must support calling the `allow` command to authorize a command. This call does not impact Docker's command line.
```
Can you wrap this to 80 chars?

thaJeztah replied:
(Other lines below as well `<3`)

thaJeztah commented on docs/extend/authorization.md:
````diff
@@ -0,0 +1,243 @@
+```
+$ docker pull centos
+…
+Authorization failed. Pull command for user 'john_doe' is denied by authorization plugin 'ACME' with message ‘[ACME] User 'john_doe' is not allowed to perform the pull command’
````
Can you wrap this example output as well? The code blocks in the documentation aren't very wide, and people will have to scroll to see the whole text otherwise

thaJeztah commented on docs/extend/authorization.md:
```diff
@@ -0,0 +1,243 @@
+
+## API schema and implementation
+
+Sample code for a typical plugin can be found here [ADD LINK]. In addition to Docker's standard plugin registration method, each plugin should implement the following two methods:
```
Looks like a link needs to be added here (`[ADD LINK]`)

dimastopel replied:
I have removed this line for now. Waiting for @moxiegirl reply on how to properly deliver the example.

thaJeztah commented on docs/reference/commandline/daemon.md:
````diff
@@ -601,6 +602,31 @@ The currently supported cluster store options are:
+`--authz-plugins=PLUGIN_ID` option.
+
+```bash
+docker daemon --authz-plugins=plugin1 --auth-plugins=plugin2,...
````
Typo in second flag (auth -> authz)

thaJeztah commented on docs/reference/commandline/daemon.md:
```diff
@@ -601,6 +602,31 @@ The currently supported cluster store options are:
+complete.
+
+For information about how to create an authorization plugin, see [authorization
+plugin](https://docs.docker.com/engine/extend/authorization/) section in the
```
Can you change this URL to `../../extend/authorization.md` ?

thaJeztah commented:
Thanks! I left some notes inline. Also, can you update https://github.com/docker/docker/blob/master/man/docker-daemon.8.md with the same information as commandline/daemon.md?

Commit:
Fixing documentation comments by @thaJeztah

Signed-off-by: Dima Stopel <dima@twistlock.com>

dimastopel commented:
@thaJeztah thanks for the thorough review!

Fixed all the comments. Added changes to `docker-daemon.8.md` as well. Wasn't sure whether to add just the command line arguments or the documentation piece as well, since I did see all the documentation from `daemon.md` copied there.  Added it eventually.

dimastopel commented:
@thaJeztah looks that changing the `plugin_api.md/#plugin-discovery` link failed the test, although it works flawlessly with `make docs`. For now removed the link since there is a link to the same page in the previous sentence. 

```
======================
Filtered (engine) Summary:

link error: (in page engine/extend/authorization.md) plugin_api.md/#plugin-discovery

	Found 1040 files
	Found 1 errors
ERRORS = 1
```

thaJeztah commented:
Thanks @dimastopel. Wondering if it works without a `/`? e.g. `plugin_api.md#plugin-discovery`

aanm commented:
@liron-l So this PR is only the authorization part or something that can replace [powerstrip](https://github.com/ClusterHQ/powerstrip)?

dimastopel commented:
@thaJeztah yes it worked! Just returned the link.

thaJeztah commented:
Cool, thanks!

thaJeztah commented:
After reading the docs, I'm still a tiny bit unsure if we should call the flag `--authz-plugins` (plural), or `--authz-plugin` (singular), given that all other flags are singular, and each flag will specify a single plugin.

@cpuguy83 @calavera final thoughts?

rhatdan commented:
Should be singular.

thaJeztah commented:
@liron-l can you change the flag to `--authz-plugin` to be more consistent with other flags? I discussed this internally, also with @moxiegirl, and if that's just a find & replace, we can merge this after! \o/

Commit:
Change authz plugin argument name

Signed-off-by: Liron Levin <liron@twistlock.com>

liron-l commented:
Thanks @thaJeztah, we've replaced all occurrences of the authz argument.

thaJeztah commented:
Thanks @liron-l, awesome!!

cyphar commented:
Awesome. Great to finally see some form of ACL support in Docker. It's been a long time coming. :smile_cat:

moxiegirl commented:
Thanks @thaJeztah ... sorry to be MIA, I'll take a look at the pr and carry any changes that might be necessary.  I'll ping @liron-l if that is actually  necessary.

thaJeztah commented:
just realized this also needs changes to the completion scripts. ping @albers @sdurrheimer if you have time to look at those? Otherwise I can see if there's a maintainer able to look at them :heart:

albers commented:
Sure, I'll take care of bash completion. Thanks for pinging me.

jfrazelle commented:
@albers you da bomb!

On Mon, Dec 14, 2015 at 5:19 AM, Harald Albers <notifications@github.com>
wrote:

> Sure, I'll take care of bash completion. Thanks for pinging me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/15365#issuecomment-164433551>.
>

sdurrheimer commented:
So do I with zsh completion.

If I understand well, there is only a single new option `--authz-plugin` for `docker daemon`, right  ?

thaJeztah commented:
@sdurrheimer correct! And to use @jfrazelle's words; you're also da bomb! thanks so much

doronp commented:
Been following this one. Awesome job. 
Just a clarification - Did you ever mean that two users will not be aware about each others stuff if need be - i.e. - two users can have 2 container with the SAME name on the same engine (otherwise the later one gets the error about name exists and hence information he should not be getting).

thaJeztah commented:
@doronp no, that still won't be possible with these changes. Container names need to be unique per daemon, however, an Authz plugin would probably be able to prevent a user from connecting to a network that is not owner by that user, or use an image that's not allowed to be used by that user. There's lots of things involved in a multi-tenant setup, and this can *help*, but won't solve all.

doronp commented:
It could also concatenate a unique ID such as user ID to the container name (modify request) and filter it out (modify response) in every response for example.
Not sure it the best path to go though.
I'd be great to see even a preliminary mapping of what is missing sometime.
WDYT?

thaJeztah commented:
@doronp yes, that would partly solve the problem, but may be quite tedious, and won't be a full solution; containers and images can be accessed by ID as well. Also think of (e.g.) volumes, and bind-mounted directories, people running docker-in-docker.

This PR offers the option to develop plugins to limit access, but for multi-tenant situations there are still lots of things to tackle. Writing up a list of things that may be needed to cover that can be a start; if you have ideas, feel free to open an issue (there may already be issues in that area)
